nabs/preprocessors.py

```diff
@@ -301,4 +303,71 @@ def daq_during_plan():
     return (yield from daq_during_plan())
 
 
+def step_size_decorator(plan):
+    """
+    Grab the last argument (number of steps), and intepret as
+    - step size if float
+    - number of steps if integer
+
+    Only works on step scans in one dimension.
+
+    Parameters
+    ----------
+    plan : plan
+        A bluesky plan that yields bluesky Msg objects.  Must be
+        a scan in one dimension, with the last argument being
+        the number of scan points / step size.
+
+    Returns
+    -------
+    step_size_plan : plan
+        The same plan as before, but modified appropriately to
+        differentiate between step size and number of steps.
+        This will be a callable generator function.
+    """
+
+    @wraps(plan)
+    def inner(*args, **kwargs):
+        if 'num' in kwargs:
+            # Currently unneeded, since daq_ascan and daq_dscan
+            # do not support num kwarg
+            n = kwargs.pop('num')
+        else:
+            # assumes (det_list, motor, start, stop, num)
+            det_list, motor, start, stop, n = args
+
+        if not isinstance(n, (numbers.Integral, float)):
+            raise TypeError("Step size / number of steps is "
+                            "neither float nor integer.")
+
+        if isinstance(n, numbers.Integral):
+            # interpret as number of steps (default)
+            result = yield from plan(*args, **kwargs)
+        elif isinstance(n, float):
```

Comment:
I think this type handling and the bit above still needs some small adjustments to handle numpy floats, if that's something that we want to allow

Reply:
Other than that I think this is gtg

Reply:
It's a good thing you pushed on this, numpy floats don't behave very nicely around zero, and I've messed with some tolerances to make these work for the test cases I could find.

Reply:
It seems like choosing an absolute tolerance value that works for everything could be difficult (imagine working in meters and scanning nanometers or microns, or the same seconds to nanoseconds with the delay stages).

I wonder if something like the following would be an improvement: "add on the `scan end` point if the final calculated point is more than 1/4 of the scan step" (where 1/4 is just a factor I arbitrarily chose). But I think the key _might_ be keeping it relative to the scan step.

Or maybe it's fine as-is and I'm overthinking it.

Reply:
That's an interesting thought.  I think that might result in inconsistent behavior if you approach the same point in large or small steps.  

For what it's worth the current implementation does use a combination of relative and absolute tolerances, they're just defined relative to the scan **points** rather than the step **size**.

Reply:
Possibly. That said, I suppose what either of us think here doesn't matter nearly as much as what the original implementation did.

The gold standard is as follows:

https://github.com/pcdshub/old-hutch-configs/blob/200f861c1fb6d66876f1f3817de9a2f24e0245b1/xpp/prod/daq.py#L1262-L1270

Reply:
.... that does solve our problem but I kinda hate it.  

`ascan(0, 2, 0.3) -> [0. , 0.3, 0.6, 0.9, 1.2, 1.5, 1.8, 2.1]`